src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -237,84 +243,112 @@ public IEnumerable<char> PotentialCommitCharacters
                 return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.None);
             }
 
-            var textChange = change.TextChange;
-            var triggerSnapshotSpan = new SnapshotSpan(triggerSnapshot, textChange.Span.ToSpan());
-            var mappedSpan = triggerSnapshotSpan.TranslateTo(subjectBuffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
-
-            using (var edit = subjectBuffer.CreateEdit(EditOptions.DefaultMinimalChange, reiteratedVersionNumber: null, editTag: null))
+            // Specifically for snippets, we use reflection to try and invoke the LanguageServerSnippetExpander's
+            // TryExpand method and determine if it succeeded or not.
+            var snippetServiceSucceeded = false;
+            if (SnippetCompletionItem.IsSnippet(roslynItem))
             {
-                edit.Replace(mappedSpan.Span, change.TextChange.NewText);
-
-                // edit.Apply() may trigger changes made by extensions.
-                // updatedCurrentSnapshot will contain changes made by Roslyn but not by other extensions.
-                var updatedCurrentSnapshot = edit.Apply();
+                var type = _languageServerSnippetExpander.GetType();
```

Comment:
@NTaylorMullen said reflection is slow, is there a way to improve performance?

Reply:
i would also cache all of this stuff up front.  and if it fails, we shouldn't add any of our snippets as there's no point if we can't expand them later.